enhancements/local-storage/automatic-detection.md

```diff
@@ -0,0 +1,157 @@
+The controller for this CR will be responsible for automatically detecting disks/devices from the available nodes and creating/managing the life cycle of `LocalVolumes` which are governed by the `AutoDetectVolumes`.s
+
+### Risks and Mitigations
+
+N/A
+
+## Design Details
+
+API scheme for `AutoDetectVolumes`:
+
+```
+type DeviceDiscoveryPolicyType string
+
+const (
+	SSDs     DeviceDiscoveryPolicyType = "ssd"
+	Disks    DeviceDiscoveryPolicyType = "disk"
+	Partions DeviceDiscoveryPolicyType = "partitions"
+)
+
+type AutoDetectVolumeList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+	Items           []AutoDetectVolume `json:"items"`
+}
+
+type AutoDetectVolume struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+	Spec              AutoDetectVolumeSpec   `json:"spec"`
+	Status            AutoDetectVolumeStatus `json:"status,omitempty"`
+}
+
+type AutoDetectVolumeSpec struct {
+	// StorageClass name to use for set of matched devices
+	StorageClassName string `json:"storageClassName"`
+	// Volume mode. Raw or with file system
+	// + optional
+	VolumeMode PersistentVolumeMode `json:"volumeMode,omitempty"`
+	// File system type
+	// +optional
+	FSType string `json:"fsType,omitempty"`
+	// Nodes on which the autoDetection must run
+	// +optional
+	NodeSelector *corev1.NodeSelector `json:"nodeSelector,omitempty"`
+	// DeviceDiscoverPolicies are the list of policies that AutoDetectVolume's controller
+	// will be looking for during detection
+	DeviceDiscoveryPolicies []DeviceDiscoveryPolicy `json:"deviceDiscoveryPolicies,omitempty"`
+}
+
+type AutoDetectVolumeStatus struct {
+	Status string `json:"status"`
+}
+
+type DeviceDiscoveryPolicy struct {
+	// Device type that should be used for auto detection. This would be one of the types supported
+	// by the local-storage operator. Initial possible types can be - crypt,raid1, raid4, raid5,
+	// raid10, multipath, disk, tape, printer, processor, worm, rom, scanner, mo-disk, changer,
+	// comm, raid, enclosure, rbc, osd, and no-lun.
+	DeviceType DeviceDiscoveryPolicyType `json:"deviceType"`
+	// A list of regular expressions that will be used to exclude certain devices
+	// For example - ["^rbd[0-9]+p?[0-9]{0,}$"]
+	// +optional
+	DeviceExclusionFilter []string `json:"deviceExclusionFilter"`
+}
+```
+
+### Test Plan
+
+- The integration tests for Local-Storage operators already exists. Those tests needs to be upgraded for testing this features.
+
+### Graduation Criteria
+
+N/A
+
+#### Examples
+
+These are generalized examples to consider, in addition to the aforementioned
+[maturity levels][maturity-levels].
+
+##### Removing a deprecated feature
+
+- None of the features are getting deprecated
+
+### Upgrade / Downgrade Strategy
+
+N/A
+
+### Version Skew Strategy
+
+N/A
+
+## Implementation History
+
+N/A
+
+## Drawbacks
+
+N/A
+
+## Alternatives
+
+N/A
```

Comment:
* Existing manual creation of `LocalVolume` CRs. With the node selector on the LocalVolume, a single CR can apply to an entire class of nodes (i.e., a machineset or a physical rack of homogeneous hardware).
  * When a machineset is defined, a corresponding LocalVolume can also be created. Why is this insufficient?
* Directly enhancing the `LocalVolume` CR to allow for autodiscovery

Reply:
> When a machineset is defined, a corresponding LocalVolume can also be created. Why is this insufficient?
UPI, direct attachment of disk to machine.

Reply:
UPI already has [many manual steps](https://docs.openshift.com/container-platform/4.3/machine_management/more-rhel-compute.html). Are we saving much?

How common is manually attaching a disk to a machine? This seems to go against the notion of machines as cattle.

Reply:
@jarrpa can you comment on the requirement for auto discovery?

Reply:
It might be worth capturing this in alternatives.

Reply:
yeah will add it.

Reply:
done